llm-service/app/services/document_storage.py

```diff
@@ -0,0 +1,79 @@
+#
+#  CLOUDERA APPLIED MACHINE LEARNING PROTOTYPE (AMP)
+#  (C) Cloudera, Inc. 2024
+#  All rights reserved.
+#
+#  Applicable Open Source License: Apache 2.0
+#
+#  NOTE: Cloudera open source products are modular software products
+#  made up of hundreds of individual components, each of which was
+#  individually copyrighted.  Each Cloudera open source product is a
+#  collective work under U.S. Copyright Law. Your license to use the
+#  collective work is as provided in your written agreement with
+#  Cloudera.  Used apart from the collective work, this file is
+#  licensed for your use pursuant to the open source license
+#  identified above.
+#
+#  This code is provided to you pursuant a written agreement with
+#  (i) Cloudera, Inc. or (ii) a third-party authorized to distribute
+#  this code. If you do not have a written agreement with Cloudera nor
+#  with an authorized and properly licensed third party, you do not
+#  have any rights to access nor to use this code.
+#
+#  Absent a written agreement with Cloudera, Inc. ("Cloudera") to the
+#  contrary, A) CLOUDERA PROVIDES THIS CODE TO YOU WITHOUT WARRANTIES OF ANY
+#  KIND; (B) CLOUDERA DISCLAIMS ANY AND ALL EXPRESS AND IMPLIED
+#  WARRANTIES WITH RESPECT TO THIS CODE, INCLUDING BUT NOT LIMITED TO
+#  IMPLIED WARRANTIES OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY AND
+#  FITNESS FOR A PARTICULAR PURPOSE; (C) CLOUDERA IS NOT LIABLE TO YOU,
+#  AND WILL NOT DEFEND, INDEMNIFY, NOR HOLD YOU HARMLESS FOR ANY CLAIMS
+#  ARISING FROM OR RELATED TO THE CODE; AND (D)WITH RESPECT TO YOUR EXERCISE
+#  OF ANY RIGHTS GRANTED TO YOU FOR THE CODE, CLOUDERA IS NOT LIABLE FOR ANY
+#  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR
+#  CONSEQUENTIAL DAMAGES INCLUDING, BUT NOT LIMITED TO, DAMAGES
+#  RELATED TO LOST REVENUE, LOST PROFITS, LOSS OF INCOME, LOSS OF
+#  BUSINESS ADVANTAGE OR UNAVAILABILITY, OR LOSS OR CORRUPTION OF
+#  DATA.
+#
+import os
+import shutil
+from abc import abstractmethod, ABC
+from pathlib import Path
+
+from . import s3
+from ..config import settings
+
+
+class DocumentStorage(ABC):
+    @abstractmethod
+    def download(self, temp_dir: str, bucket_name: str, document_key: str, original_filename: str) -> Path:
+        """
+        Copy file from storage into the temp directory
+        """
+
+    @staticmethod
+    def from_environment() -> "DocumentStorage":
+        # todo: move this to config, remove bucket_name from download api signature
+        bucket = os.environ.get("S3_RAG_DOCUMENT_BUCKET")
+        if bucket:
+            return S3DocumentStorage()
+        else:
+            return FileSystemDocumentStorage()
+
+class S3DocumentStorage(DocumentStorage):
+    def download(self, temp_dir: str, bucket_name: str, document_key: str, original_filename: str) -> Path:
+        """
+        Download document from S3
+        """
+        return s3.download(temp_dir, bucket_name, document_key, original_filename)
```

Comment:
nit: maybe we should just copy the content of `s3.download`? Instead of having two ways to deal with S3, since we want to incentivize `DocumentStorage`